pkg/clusteragent/autoscaling/workload/controller.go

```diff
@@ -0,0 +1,333 @@
+}
+
+// Synchronize DatadogPodAutoscaler state between internal store and Kubernetes objects
+// Make sure any `return` has the proper store Unlock
+func (c *Controller) syncPodAutoscaler(ctx context.Context, key, ns, name string, podAutoscaler *datadoghq.DatadogPodAutoscaler) (autoscaling.ProcessResult, error) {
+	podAutoscalerInternal, podAutoscalerInternalFound := c.store.LockRead(key, true)
+
+	// Object is missing from our store
+	if !podAutoscalerInternalFound {
+		if podAutoscaler != nil {
+			// If we don't have an instance locally, we create it. Deletion is handled through setting the `Deleted` flag
+			c.store.UnlockSet(key, model.NewPodAutoscalerInternal(key, podAutoscaler), c.ID)
+		} else {
+			// If podAutoscaler == nil, both objects are nil, nothing to do
+			c.store.Unlock(key)
+		}
+
+		return autoscaling.NoRequeue, nil
+	}
+
+	if podAutoscaler == nil {
+		// Object is not present in Kubernetes
+		// If flagged for deletetion, we just need to clear up our store (deletion complete)
+		// Also if object was not owned by remote config, we also need to delete it (deleted by user)
+		if podAutoscalerInternal.Deleted || podAutoscalerInternal.Spec.Owner != datadoghq.DatadogPodAutoscalerRemoteOwner {
+			c.store.UnlockDelete(key, c.ID)
+			return autoscaling.NoRequeue, nil
+		}
+
+		// Object is not flagged for deletion and owned by remote config, we need to create it in Kubernetes
+		err := c.createPodAutoscaler(ctx, podAutoscalerInternal)
+
+		c.store.Unlock(key)
+		return autoscaling.Requeue, err
+	}
+
+	// Object is present in both our store and Kubernetes, we need to sync depending on ownership.
+	// Implement info sync based on ownership.
+	if podAutoscaler.Spec.Owner == datadoghq.DatadogPodAutoscalerRemoteOwner {
+		// First implement deletion logic, as if it's a deletion, we don't need to update the object.
+		// Deletion can only happen if the object is owned by remote config.
+		if podAutoscalerInternal.Deleted {
+			err := c.deletePodAutoscaler(ns, name)
+			// In case of not found, it means the object is gone but informer cache is not updated yet, we can safely delete it from our store
+			if err != nil && errors.IsNotFound(err) {
+				c.store.UnlockDelete(key, c.ID)
+				return autoscaling.NoRequeue, nil
+			}
+
+			// In all other cases, we requeue and wait for the object to be deleted from store with next reconcile
+			c.store.Unlock(key)
+			return autoscaling.Requeue, err
+		}
+
+		// If the object is owned by remote config and newer, we need to update the spec in Kubernetes
+		// If Kubernetes is newer, we wait for RC to update the object in our internal store.
+		if *podAutoscalerInternal.SettingsVersion > podAutoscaler.Spec.RemoteVersion {
+			podAutoscaler.Spec = podAutoscalerInternal.Spec
+			err := c.updatePodAutoscalerSpec(ctx, podAutoscalerInternal, podAutoscaler)
+
+			// If update succeeded, we requeue the object to handle status update
+			if err == nil {
+				c.Workqueue.AddRateLimited(key)
```

Comment:
When returning `autoscaling.Requeue` aren't we already already calling `AddRateLimited` ?